hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -630,7 +642,27 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
     val rightKeyFieldIdx = rightRVDType.kFieldIdx
     val leftValueFieldIdx = leftRVDType.valueFieldIdx
     val rightValueFieldIdx = rightRVDType.valueFieldIdx
-    val newRowPType = PType.canonical(newRowType).setRequired(true).asInstanceOf[PStruct]
+
+    val (lkT, lvT, rvT) = joinType match {
+      case "inner" =>
+        val keyTypeFields = castFieldRequiredeness(leftRVDType.kType, true)
+        (keyTypeFields, leftRVDType.valueType.fields, rightRVDType.valueType.fields)
+      case "left" =>
+        val rValueTypeFields = castFieldRequiredeness(rightRVDType.valueType, false)
+        (leftRVDType.kType.fields, leftRVDType.valueType.fields, rValueTypeFields)
+      case "right" =>
+        val lValueTypeFields = castFieldRequiredeness(leftRVDType.valueType, false)
+        (leftRVDType.kType.fields, lValueTypeFields, rightRVDType.valueType.fields)
+      case "outer" | "zip" =>
+        val keyTypeFields = castFieldRequiredeness(leftRVDType.kType, false)
```

Comment:
this can actually be the union of requiredeness of keys.

Reply:
> (they all do, but you can do a setRequired to the left in left join, right in right join, && in inner)

Yes, this is correct. Left needs the left type, right should have the same field names but the right's ptype, and outer/zip get the union.

Reply:
I added a test. We did not have one that could correctly identify requiredeness differences between the two tables during left/right